modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/contributor/extender/internal/servlet/taglib/IconContributorDynamicInclude.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.icons.admin.web.contributor.extender.internal.servlet.taglib;
+
+import com.liferay.frontend.icons.admin.web.contributor.extender.IconResource;
+import com.liferay.frontend.icons.admin.web.contributor.extender.IconResourcesContributor;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.servlet.taglib.DynamicInclude;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * @author Bryce Osterhaus
+ */
+@Component(immediate = true, service = DynamicInclude.class)
+public class IconContributorDynamicInclude implements DynamicInclude {
+
+	@Override
+	public void include(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String key)
+		throws IOException {
+
+		PrintWriter printWriter = httpServletResponse.getWriter();
+
+		printWriter.write("<svg style=\"display:none;\">");
+
+		Lock lock = _readWriteLock.readLock();
+
+		lock.lock();
+
+		try {
+			Set<Map.Entry<String, List<IconResource>>> entries =
+				_iconResourcesMap.entrySet();
+
+			for (Map.Entry<String, List<IconResource>> entry : entries) {
+				List<IconResource> iconResources = entry.getValue();
+
+				IconResource iconResource = iconResources.get(0);
+
+				printWriter.print(iconResource.getSVGContent());
+			}
+		}
+		finally {
+			lock.unlock();
+		}
+
+		printWriter.write("</svg>");
+	}
```

Comment:
@izaera I got this code from your old PR. I noticed it is causing some weird jank in the page when loading the page, its almost as if the css doesn't get loaded. You will see it in the page changes. I tried removing this code chunk and I no longer get jank. Is there something special I need to do? See gif below for example of css/jank loading.

![Kapture 2021-08-19 at 15 23 45](https://user-images.githubusercontent.com/6843530/130152800-34034ebc-9d76-4393-9503-e6c9a9a0c4de.gif)

Reply:
:shrug: I don't believe it is related to the code itself, but some interaction related to how the browser renders the page. 

Maybe we are writing a lot of `<svg>`s in the `<head>` before the CSS `<link>`s and that causes a browser delay, or something like that... :thinking:

Reply:
I've suffered too and it worries me, as it invalidates this approach until we fix it. Could be that the lock is adding contingency? I can help investigating it if you want.

Reply:
I was going to test it today, but had to prepare another PR, so I'll begin now and will probably say something by tomorrow :crossed_fingers: